src/backends/mod.rs

```diff
@@ -35,10 +35,205 @@
+}
+
+/// A builder interface for instantiating backends.
+pub struct TracerBuilder {
+    user_backend: Option<BackendKind>,  // User-requested backend choice.
+    auto_backend: BackendKind,          // Auto-chosen backend if `user_backend` is `None`.
+    config: BackendConfig,
+}
+
+impl TracerBuilder {
+    /// Create a new TracerBuilder with default options.
+    pub fn new() -> Self {
+        Self {
+            user_backend: None,
+            auto_backend: Self::default_platform_backend(),
+            config: BackendConfig {
+                perf_aux_bufsize: None,
+                perf_data_bufsize: None,
+                new_trace_bufsize: None,
+            }
+        }
+    }
+
+    /// Overide auto-default backend selection.
+    pub fn backend(mut self, backend: BackendKind) -> Self {
+        self.user_backend = Some(backend);
+        self
+    }
+
+    /// Set the Linux Perf data buffer size (in pages).
+    /// Applies to only backends using Linux perf.
+    pub fn perf_data_bufsize(mut self, size: usize) -> Self {
+        self.config.perf_data_bufsize = Some(size);
+        self
+    }
+
+    /// Set the Linux Perf AUX buffer size (in pages).
+    /// Applies to only backends using Linux perf.
+    pub fn perf_aux_bufsize(mut self, size: usize) -> Self {
+        self.config.perf_aux_bufsize = Some(size);
+        self
+    }
+
+    /// Set the initial trace storage buffer size (in bytes).
+    /// Applies to all backends.
+    pub fn new_trace_bufsize(mut self, size: usize) -> Self {
+        self.config.new_trace_bufsize = Some(size);
+        self
+    }
+
+    /// Query which backend will be used.
+    /// Useful if you've allowed the builder to choose an auto-default, but wish to know which.
+    pub fn backend_kind(&self) -> &BackendKind {
+        if let Some(ref backend) = self.user_backend {
+            &backend
+        } else {
+            &self.auto_backend
+        }
+    }
+
+    // Finds a suitable `BackendKind` for the current hardware/OS.
+    fn default_platform_backend() -> BackendKind {
+        let tr_kinds = vec![BackendKind::PerfPT, BackendKind::Dummy];
+        for kind in tr_kinds {
+            if Self::match_platform(&kind).is_ok() {
+                return kind;
+            }
+        }
+        // The Dummy backend should always be usable.
+        unreachable!();
+    }
+
+    /// Returns `Ok` if the tracer backend should work on the current hardware/OS.
+    fn match_platform(kind: &BackendKind) -> Result<(), HWTracerError> {
+        match kind {
+            BackendKind::Dummy => Ok(()),
+            BackendKind::PerfPT => {
+                #[cfg(not(perf_pt))]
+                return Err(HWTracerError::BackendUnavailable(BackendKind::PerfPT));
+                #[cfg(perf_pt)] {
+                    if !Self::pt_supported() {
+                        return Err(HWTracerError::NoHWSupport("Intel PT not supported by CPU".into()));
+                    }
+                    return Ok(());
```

Comment:
Can just be `Ok(())`.

Reply:
Maybe. Conditional compilation often gets in the way of expression evaluation. I've been meaning to make a bug report. Will try.

Reply:
Yeah, this one actually works.

Reply:
Here's an example of the kind of stuff I've noticed:
```
fn num() -> usize {
    #[cfg(test)]
    return 1;
    #[cfg(not(test))]
    2
}
```

Without the `return` this fails to compile. Yet the second `#[cfg]` and the following line should simply disappear. Weird.

Reply:
It's not really surprising: it's a sort-of limitation of the parsing. You might think you could write:

```rust
fn num() -> usize {
    #[cfg(test)]
    1
    #[cfg(not(test))]
    2
}
```

but (I imagine) that's unparseable even though "semantically" it makes sense.

Reply:
Yes, that's exactly what I thought I could write. Blame C.

```
long long int num() {
    #ifdef TEST
    return 1;
    #else
    return 2;
}
```

Thanks to the preprocessor the parser doesn't even see one arm of the compile time constant. I assumed Rust would do the same, but now I see why it cannot.